File: apps/backend/src/modules/lawyer/application/commands/register-lawyer/register-lawyer.handler.ts
import { CommandHandler, ICommandHandler } from '@nestjs/cqrs';
import { Inject } from '@nestjs/common';
import { RegisterLawyerCommand } from './register-lawyer.command';
import { Result } from '@shared/domain/result';
import { Lawyer } from '../../../domain/entities';
import { LicenseNumber, Experience } from '../../../domain/value-objects';
import { ILawyerRepository } from '../../../domain/repositories/lawyer.repository.interface';

export type RegisterLawyerResult = Result<{ lawyerId: string }>;

/**
 * RegisterLawyerCommandHandler
 *
 * Handles lawyer registration business logic
 */
@CommandHandler(RegisterLawyerCommand)
export class RegisterLawyerCommandHandler
  implements ICommandHandler<RegisterLawyerCommand, RegisterLawyerResult>
{
  constructor(
    @Inject('ILawyerRepository')
    private readonly lawyerRepository: ILawyerRepository,
  ) {}

  async execute(command: RegisterLawyerCommand): Promise<RegisterLawyerResult> {
    // 1. Validate and create LicenseNumber value object
    const licenseNumberOrError = LicenseNumber.create(command.licenseNumber);
    if (licenseNumberOrError.isFailure) {
      return Result.fail(licenseNumberOrError.error);
    }
    const licenseNumber = licenseNumberOrError.value;

    // 2. Check if license number already exists
    const exists = await this.lawyerRepository.existsByLicenseNumber(
      licenseNumber.value,
    );
    if (exists) {
      return Result.fail('License number already registered');
    }

    // 3. Check if user already has lawyer profile
    const existingLawyer = await this.lawyerRepository.findByUserId(
      command.userId,
    );
    if (existingLawyer) {
      return Result.fail('User already has a lawyer profile');
    }

    // 4. Create Experience value object
    const experienceOrError = Experience.create(command.experienceYears);
    if (experienceOrError.isFailure) {
      return Result.fail(experienceOrError.error);
    }
    const experience = experienceOrError.value;

    // 5. Generate ID
    const lawyerId = await this.lawyerRepository.nextId();

    // 6. Create Lawyer entity
    const lawyerOrError = Lawyer.create(
      lawyerId,
      command.userId,
      licenseNumber,
      command.specializations,
      experience,
      command.bio,
      command.education,
      command.hourlyRate,
    );

    if (lawyerOrError.isFailure) {
      return Result.fail(lawyerOrError.error);
    }

    const lawyer = lawyerOrError.value;

    // 7. Save lawyer
    await this.lawyerRepository.save(lawyer);

    // 8. Return success result
    return Result.ok({ lawyerId: lawyer.id });
  }
}

File: apps/backend-python/app/modules/lawyer/application/commands/register_lawyer_handler.py
"""
RegisterLawyerHandler

Handler для регистрации нового юриста.
"""

from typing import Tuple

from app.core.domain.result import Result
from ...domain.entities.lawyer import Lawyer
from ...domain.repositories.lawyer_repository import ILawyerRepository
from ...domain.services.lawyer_verification_service import LawyerVerificationService
from ...domain.value_objects.experience import Experience
from ...domain.value_objects.price import Price
from ...domain.value_objects.specialization import Specialization
from ..dtos.lawyer_dto import LawyerDTO
from .register_lawyer import RegisterLawyerCommand


class RegisterLawyerHandler:
    """
    Handler для команды RegisterLawyerCommand.

    Процесс:
    1. Проверяет, что user_id не зарегистрирован как юрист
    2. Валидирует данные через VerificationService
    3. Создает Value Objects (Specialization, Experience, Price)
    4. Создает Lawyer aggregate через factory method
    5. Сохраняет в репозиторий
    6. Возвращает LawyerDTO

    Dependencies:
        lawyer_repository: Репозиторий юристов
    """

    def __init__(self, lawyer_repository: ILawyerRepository) -> None:
        """
        Инициализирует handler.

        Args:
            lawyer_repository: Репозиторий юристов
        """
        self.lawyer_repository = lawyer_repository

    async def handle(
        self, command: RegisterLawyerCommand
    ) -> Result[LawyerDTO]:
        """
        Обрабатывает команду регистрации юриста.

        Args:
            command: Команда с данными

        Returns:
            Result с LawyerDTO или ошибкой
        """
        # 1. Проверяем, что пользователь еще не зарегистрирован как юрист
        existing_lawyer = await self.lawyer_repository.find_by_user_id(command.user_id)
        if existing_lawyer:
            return Result.fail("User is already registered as a lawyer")

        # 2. Валидация через Domain Service
        eligibility_result = LawyerVerificationService.is_eligible_for_verification(
            license_number=command.license_number,
            education=command.education,
            about=command.about,
        )

        if eligibility_result.is_failure:
            # Форматируем ошибки для пользователя
            errors = eligibility_result.error
            error_messages = []
            for field, messages in errors.items():
                error_messages.extend(messages)
            return Result.fail("; ".join(error_messages))

        # 3. Создаем Value Objects
        try:
            # Специализации
            specializations = []
            for spec_str in command.specializations:
                spec = Specialization(spec_str)
                specializations.append(spec)

            # Опыт
            experience = Experience(command.experience_years)

            # Цена
            price = Price(command.price_per_consultation)

        except ValueError as e:
            return Result.fail(str(e))

        # 4. Создаем Lawyer aggregate
        lawyer_result = Lawyer.create(
            user_id=command.user_id,
            specializations=specializations,
            experience=experience,
            price_per_consultation=price,
            license_number=command.license_number,
            education=command.education,
            about=command.about,
            location=command.location,
            languages=command.languages,
        )

        if lawyer_result.is_failure:
            return Result.fail(lawyer_result.error)

        lawyer = lawyer_result.value

        # 5. Сохраняем в репозиторий
        await self.lawyer_repository.save(lawyer)

        # 6. Возвращаем DTO
        lawyer_dto = LawyerDTO.from_entity(lawyer)
        return Result.ok(lawyer_dto)
